org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/codegen/BranchLabel.java

```diff
@@ -85,6 +85,29 @@ public void becomeDelegateFor(BranchLabel otherLabel) {
 	// other label is delegating to receiver from now on
 	otherLabel.delegate = this;
 
+	if (this.targetStackDepth == -1) {
+		if (this.codeStream.stackDepth < 0) {
+			this.codeStream.classFile.referenceBinding.scope.problemReporter()
+					.operandStackSizeInappropriate(this.codeStream.classFile.referenceBinding.scope.referenceContext);
+			this.codeStream.stackDepth = 0; // FWIW
+			this.codeStream.operandStack.clear();
+		}
+		this.targetStackDepth = this.codeStream.stackDepth;
+		this.operandStack = this.codeStream.operandStack.copy();
+		// TODO: check that operand stack contents slot count matches targetStackDepth
+	} else {
+		// Stack depth known at label having encountered a previous branch and/or having fallen through to label
+		if (this.targetStackDepth != this.codeStream.stackDepth) {
+			this.codeStream.classFile.referenceBinding.scope.problemReporter().operandStackSizeInappropriate(
+					this.codeStream.classFile.referenceBinding.scope.referenceContext);
+			if (this.targetStackDepth < this.codeStream.stackDepth) {
+				this.targetStackDepth = this.codeStream.stackDepth; // FWIW, pick the higher water mark.
+				this.operandStack = this.codeStream.operandStack.copy();
+			}
+		}
+		// TODO: check that operand stacks match.
+	}
+
 	// all existing forward refs to other label are inlined into current label
```

Comment:
A `BranchLabel`s `targetStackDepth` get set under two circumstances. When a branch is issued on the label (`org.eclipse.jdt.internal.compiler.codegen.BranchLabel.branch()`) or when the `BranchLabel` is placed (`org.eclipse.jdt.internal.compiler.codegen.BranchLabel.place()`) such that the previous instruction fall through to the instruction with the label. 

In the defect scenario, `org.eclipse.jdt.internal.compiler.codegen.CodeStream.goto_(BranchLabel)` via `org.eclipse.jdt.internal.compiler.codegen.CodeStream.inlineForwardReferencesFromLabelsTargeting(BranchLabel, int)` finds out that it is a chained goto situation and arranges the target label to delegate to the one of the pre-existing labels and short circuits the emission of the `goto` - As the call to `org.eclipse.jdt.internal.compiler.codegen.BranchLabel.branch()` from `org.eclipse.jdt.internal.compiler.codegen.CodeStream.goto_(BranchLabel)` is skipped, we fail to the update the stack depth at the target of the goto. 

The fix is to update `org.eclipse.jdt.internal.compiler.codegen.BranchLabel.becomeDelegateFor(BranchLabel)` and propagate the target stack depth there